src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/SignatureHelper.cs

```diff
@@ -46,46 +44,95 @@ internal static BlobBuilder MethodSignatureEncoder(ModuleBuilderImpl _module, Ty
             {
                 foreach (Type parameter in parameters)
                 {
-                    WriteSignatureTypeForReflectionType(parEncoder.AddParameter().Type(), parameter);
+                    WriteSignatureTypeForReflectionType(parEncoder.AddParameter().Type(), parameter, module);
                 }
             }
 
             return methodSignature;
         }
 
-        private static void WriteSignatureTypeForReflectionType(SignatureTypeEncoder signature, Type type)
+        private static void WriteSignatureTypeForReflectionType(SignatureTypeEncoder signature, Type type, ModuleBuilderImpl module)
         {
-            // We need to translate from Reflection.Type to SignatureTypeEncoder. Most common types for proof of concept. More types will be added.
-            // TODO: This switch should be done by comparing Type objects, without fetching FullName.
-            switch (type.FullName)
+            // We need to translate from Reflection.Type to SignatureTypeEncoder.
+            if (Enum.TryParse(type.Name, out CoreTypeId typeId) && type.Equals(module.GetTypeFromCoreAssembly(typeId)))
```

Comment:
We should not be parsing the names to do this mapping. We should be comparing type objects instead. (You can add a method next to GetTypeFromCoreAssembly that will go through all core types to make this efficient.)